include/ginkgo/core/matrix/coo.hpp

```diff
@@ -258,6 +258,31 @@ class Coo : public EnableLinOp<Coo<ValueType, IndexType>>,
         return this;
     }
 
+    /**
+     * Creates a constant (immutable) Coo matrix from a set of constant arrays.
+     *
+     * @param exec  the executor to create the matrix on
+     * @param size  the dimensions of the matrix
+     * @param values  the value array of the matrix
+     * @param col_idxs  the column index array of the matrix
+     * @param row_ptrs  the row index array of the matrix
+     * @returns A smart pointer to the constant matrix wrapping the input arrays
+     *          (if they reside on the same executor as the matrix) or a copy of
+     *          these arrays on the correct executor.
+     */
+    static std::unique_ptr<const Coo> create_const(
```

Comment:
Should `create_const` be abstracted into an interface, the same way it's done for `create`? (Same goes for all other matrix types)

Reply:
This might require quite a lot of plumbing code due to the different sets of parameters the different matrix types require - variadic parameter set, conditionally transforming parameters via `array_const_cast`. Do you think that would be worth the effort?

Reply:
I'm not sure if I follow here. I had the exact same setup as with `EnableCreateMethod` in mind. For example, the interface could be just
```c++
template <typename ConcreteType>
class EnableCreateConstMethod {
public:
    template <typename... Args>
    static std::unique_ptr<const ConcreteType> create_const(Args&&... args)
    {
        return std::unique_ptr<const ConcreteType>(
            new ConcreteType(std::forward<Args>(args)...));
    }
};
```
and then the same implementation for the `create_const` would be moved to the constructor, e.g.
```c++
Coo(std::shared_ptr<const Executor> exec, const dim<2>& size,
    Array<const ValueType>&& values, Array<const IndexType>&& col_idxs,
    Array<const IndexType>&& row_idxs)
: Coo(std::move(exec), size, gko::detail::array_const_cast(std::move(values)),
        gko::detail::array_const_cast(std::move(col_idxs)),
        gko::detail::array_const_cast(std::move(row_idxs)))
{}
```
I don't think that would be much of an effort, except that this would need to be changed for every matrix format.

Reply:
Ok, I get the point. Before this addition, the templated constructor would not work with const arrays, so you would get a compiler error if you tried that before (at least that's what I hope). I haven't tested it, but with the changes to `Array` would the templated constructor now compile with const arrays? If so, then we already have that problem.

Reply:
Just checked, and even with the changes to `Array` you can't use the templated constructor with const arrays. So then we don't add the interface, at least not until we have a neat way to introduce it.

Reply:
This brings up some good points - maybe we should instead introduce a `ConstArrayView`-like type that doesn't have any ownership semantics or assignment operators, and use that instead of `Array<const>`? It would look the same for users, but avoid any workarounds to make Array<const> behave consistently?